### PR TITLE
feat(expr): allow function to return runtime error

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- allow function to return runtime error
- `Column::index()` returns `Option` instead of panic
- test throwing error in `get(Array(T0), Int16) :: T0`

Tracked in https://github.com/datafuselabs/databend/issues/6547

### CODE REVIEW

**Review by leiysky — APPROVED** [review]

**Review by Xuanwo — APPROVED** [review]
Please make clippy happy

File: common/expression/src/function.rs [hunk 1]
```diff
@@ -477,157 +494,160 @@ fn erase_calc_domain_generic_2_arg<I1: ArgType, I2: ArgType, O: ArgType>(
...
-    move |_args, generics| func(generics).upcast()
+    func: impl Fn(&GenericMap) -> Result<Value<O>, String>,
+) -> impl Fn(&[ValueRef<AnyType>], &GenericMap) -> Result<Value<AnyType>, String> {
+    move |_args, generics| func(generics).map(Value::upcast)
 }
 
 fn erase_function_generic_1_arg<I1: ArgType, O: ArgType>(
-    func: impl Fn(ValueRef<I1>, &GenericMap) -> Value<O>,
-) -> impl Fn(&[ValueRef<AnyType>], &GenericMap) -> Value<AnyType> {
+    func: impl Fn(ValueRef<I1>, &GenericMap) -> Result<Value<O>, String>,
```
Comment (sundy-li) [inline common/expression/src/function.rs:503]: If we always return `Ok`, will this prevent the Auto-Vectorized optimization?

Comment (andylokandy) [reply common/expression/src/function.rs:503]: There are four ways to register a function overload: `register_n_arg`, `register_with_writer_n_arg`, `register_n_arg_core`, and `register_factory`. They have different features as described in the table:

|                            | Auto Vectorization | Auto Null Passthrough | Auto Downcast | Access Output Column Builder | Throw Runtime Error | Varidic |
|----------------------------|--------------------|-----------------------|---------------|------------------------------|---------------------|---------|
| register_n_arg             | ✅                  | ✅                     | ✅             | ❌                            | ❌                   | ❌       |
| register_with_writer_n_arg | ✅                  | ✅                     | ✅             | ✅                            | ✅                   | ❌       |
| register_n_arg_core        | ❌                  | ❌                     | ❌             | ✅                            | ✅                   | ❌       |
| register_factory           | ❌                  | ❌                     | ❌             | ✅                            | ✅                   | ✅       |

Comment (sundy-li) [reply common/expression/src/function.rs:503]: Ok, let's keep this table in mind.

Comment (Xuanwo) [reply common/expression/src/function.rs:503]: > They have different features as described in the table:

Do we have this table in docs or code comments?

**Pull request merged**